flypipe/node.py

```diff
@@ -369,8 +328,49 @@ def _get_consolidated_output_schema(cls, output_schema, output_columns):
             schema = None
         return schema
 
+    def process_transformation_with_cache(
+            self, runnable_node, run_context: RunContext
+    ):
+
+        result = None
+        node_transformation = runnable_node["transformation"]
+        ran_transformation = False
+
+        if runnable_node["status"] == RunStatus.CACHED and \
+                runnable_node["node_run_context"].cache_context.exists_cache_to_load:
+            result = runnable_node["node_run_context"].cache_context.read()
+        else:
+            ran_transformation = True
+            dependency_values = runnable_node["transformation"].get_node_inputs(
+                run_context
+            )
+
+            result = node_transformation.process_transformation(
+                run_context.spark,
+                runnable_node["output_columns"],
+                runnable_node["node_run_context"],
+                **dependency_values,
+            )
+
+        run_context.update_node_results(
+            node_transformation.key,
+            result,
+            schema=self._get_consolidated_output_schema(
+                node_transformation.output_schema,
+                runnable_node["output_columns"],
+            ),
+        )
+
+        if ran_transformation:
+            # If cache exists, and the transformation has run, then save its cache
+            runnable_node["node_run_context"].cache_context.write(
+                run_context.node_results[node_transformation.key]

```

Comment:
Hmm are you sure it's fine to save the result into the cache after columns have been selected? Suppose there's a cached node that returns columns [c1, c2, c3] but only has requested columns [c1, c2] in a graph. Then you use this node in another graph which requests all 3 columns. The cache is per node and not per graph and therefore c3 will be lost.

Reply:
it is fine as before that

```
        run_context.update_node_results(
            node_transformation.key,
            result,
            schema=self._get_consolidated_output_schema(
                node_transformation.output_schema,
                runnable_node["output_columns"],
            ),
        )
```

it saves the dataframe with the schema. so it will use the provided schema of the node to write the cache. if not provided (declared in the node), it will use the requested columns from the graph.

In the documentation we can state that it will save the cache with the declared output schema, if not provided, it will save the cache with the requested columns from the graph

Reply:
👍